File: lab2/fake_device.py
import os
import sys

inconnection = "/tmp/test_connection_dev"
outconnection = "/tmp/test_connection_driv"
try:
	os.mkfifo(inconnection)
	os.mkfifo(outconnection)
except OSError:
	pass
infifo = os.open(inconnection, os.O_RDONLY)
outfifo = os.open(outconnection, os.O_WRONLY)
disk = []
	
exits = False

def process_open_request():
	fifo.close()
	fifo = open(connection, "w")
	fifo.write("OPEN_ACK")
	fifo.close()
	print("Open Device Processed")
	fifo = open(connection, "r")
	
def process_release_request():
	fifo.close()
	fifo = open(connection, "w")
	fifo.write("REL_ACK")
	fifo.close()
	print("Close Device Processed")
	fifo = open(connection, "r")

def process_read_request():
	fifo.close()
	fifo = open(connection, "w")
	for i in disk:
		fifo.write(i)
	fifo.write("READ_COMP")
	fifo.close()
	print("Read Request Processed")
	fifo.open(connection, "r")

def process_write_request(fifo):
	fifo = open(connection, "r")
	for line in fifo:
		disk.append(line)
	fifo.close()
	print("disk contents:\n")
	for i in disk:
		print(i)
	fifo = open(connection, "r+")
	

while(exits == False):
	line = os.read(infifo, 52)
	"""if line == "release":
		process_release_request()
	elif line == "open":
		process_open_request()
	elif line == "read":
		process_read_request()"""
	if (True):
		print("WRITE REQUEST!")
		os.write(outfifo, bytes("WRITE_ACK", 'UTF-8'))
		print("Write Request Processed")
	else:
		infifo.close()
		exits = True

File: lab2/fake_device.c
#include <stdio.h>
#include <unistd.h>
#include <fcntl.h>
#include <string.h>
#include <sys/stat.h>
static char DISK_BUFFER[1024];
static char INT_BUFFER[512];
static char DISK_STAT[1];

int main() {
	int fifIn;
	int fifOut;
	int eof = 0;
	char * diskFifo = "/tmp/virtual_dev_in";
	char * drivFifo = "/tmp/virtual_dev_out";
	mkfifo(diskFifo, 0666);
	mkfifo(drivFifo, 0666);
	for (int i = 0; i < sizeof(DISK_STAT); i++)
		DISK_STAT[i] = 0;
	
	fifIn = open(diskFifo, O_RDONLY | O_CREAT, 0666);
	fifOut = open(drivFifo, O_WRONLY | O_CREAT, 0666);
	
	while(1) {
		read(fifIn, INT_BUFFER, 1024);
		if(strcmp(INT_BUFFER, "WRITE_REQ") == 0 && DISK_STAT[0] == 0) {
			printf("WRITE REQUEST\n");
			DISK_STAT[0] = 1;
			write(fifOut, "WRITE_ACK", sizeof("WRITE_ACK"));
		}
		else if(DISK_STAT[0] == 1 && strcmp(INT_BUFFER, "WRITE_REQ") != 0) {
			int i;
			for(i = 0; i < sizeof(INT_BUFFER) && INT_BUFFER[i] != '\0'; i++) {
				DISK_BUFFER[eof] = INT_BUFFER[i];
				eof++;
			}
			DISK_STAT[0] = 0;
			printf("%s", DISK_BUFFER);
		}  
	}
	/*close(fifIn);
	close(fifOut);
	exit(0);*/
}
